### PR TITLE
Random or Sequential snippets. Overloaded +. Removed '\n' from text by default for clean.

### PR DESCRIPTION
**Pull request by diptanil-roy**
1. random_snippet returns random or sequential snippets based on input. 
2. Overloaded +. So, Book1 + Book2 possible now.
3. Removed all '\n' characters in clean by default.

### CODE REVIEW

File: src/book.py [hunk 1]
```diff
@@ -82,7 +82,37 @@ def __init__(self, filepath=None, rawtext=None, encoding='utf-8', file_format='s
                 else:
                     self._title = None
         else:
-            self._title = None
+            self._title = title #FIXME IN MASTER
+            
+    def __add__(self, other):
```
Comment (emapple) [inline src/book.py:89]: A couple things for docstrings --

- always use three double quotes, not single quotes, for docstrings
- the first line should be a short summary, one line maximum

For more, see https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

File: src/book.py [hunk 2]
```diff
@@ -82,7 +82,37 @@ def __init__(self, filepath=None, rawtext=None, encoding='utf-8', file_format='s
...
+        if (title1 is None and title2 is not None):
+            title = title2
+        elif (title1 is not None and title2 is None):
+            title = title1
+        elif (title1 != title2):
+            print ("The authors are not the same. Assigning the title of the 1st Book to the resultant bookobject")
+            title = title1
+        else:
+            title = title1
+        return BookText(rawtext=self._text+other._text, author=author, title=title, meta=None)
```
Comment (emapple) [inline src/book.py]:
```suggestion
        return BookText(rawtext=self._text+' '+other._text, author=author, title=title, meta=None) 
```
Let's add a space between subsequent texts for readability.

File: src/book.py [hunk 3]
```diff
@@ -115,7 +146,12 @@ def clean(self, lemmatize=True, deromanize=False, lemma_pos='v', inplace=False):
                     WNLemma.lemmatize(token, pos='n'), pos='v'), pos='a')
                     for token in tokens]
             cleaned = (" ").join(lemmatized)
-
+        else:
+            # Added this bit just to clean up unnecessary spaces in the text.
+            # The Clean Function thus returns a text with no extra spaces.
+            tokens = word_tokenize(cleaned)
+            cleaned = (" ").join(tokens)
+
```
Comment (emapple) [inline src/book.py]: Technically this will introduce spaces around punctuation that might not otherwise exist. If your goal is to replace any grouping of spaces with a single space, I propose this instead:
```suggestion
        else:
            # Added this bit just to clean up unnecessary spaces in the text.
            # The Clean Function thus returns a text with no extra spaces.
            cleaned = re.sub('\s+', ' ', cleaned)
            
```

File: src/book.py [hunk 4]
```diff
@@ -125,7 +161,7 @@ def clean(self, lemmatize=True, deromanize=False, lemma_pos='v', inplace=False):
         if inplace:
             self._text = cleaned
         else:
-            return BookText(filepath=None, rawtext=cleaned, author=self.author, title=self.title, meta=self.meta)
+            return BookText(filepath=None, rawtext=cleaned, author=self._author, title=self._title, meta=self._meta)
```
Comment (emapple) [inline src/book.py]: Doesn't it work either way? I haven't run into any issues here.

**Review by emapple — CHANGES_REQUESTED** [review]
These are all good changes (I quite like your overloaded + operator).

I've left some specific comments throughout, but here are my comments about the `random_snippet` method.

1) This probably shouldn't be a separate function. It should be an update of `snippet`, since now we have redundant functionality and a lot of copy-pasted text.

If you really want to keep the simpler functionality of `snippet`, it should be a wrapper function that calls `random_snippet` with a default of one group. Ideally, it would also return a BookText object and not an array of BookText objects so as not to break existing functionality. Alternatively, just have the one function `snippet`, with an if/else statement at the end so that if `n_groups=1`, it returns a BookText object instead of an array of BookText objects.

2) This isn't doing what I expect it to right now. For example,
```
bt = BookText(rawtext='This is a test with seven words.')
bt.random_snippet('word', n_on=3, randomized=True)[0].text
out:   'is test words '
```
is returning 3 words in random order, whereas I expect it to return 3 sequential words in a random starting point. This would be consistent with the behavior of `snippet`, and I think it's how we should keep it. Three random words should be `n_groups=3, n_on=1, randomized=True`, not `n_groups=1, n_on=3, randomized=True`.

Perhaps the solution is to allow `start_point=0` to indicate a random start point, and set the default to `start_point=0` and `randomized=False`. This will reproduce the behavior of `snippet` (and the defaults should be set so that this is all backwards compatible. This also means `n_on` should be changed from a keyword argument to a positional argument). I know I will far more often want a random snippet of contiguous text than of non-contiguous text.

3) It's really hard for me to follow your code right now because way too much is being done in a single line. Please break up the long lines by creating variables that you use later. For example, rather than using `1 + (iteration)*n_on` over and over, just create a variable 
```
end = 1 + (iteration)*n_on
```
and use the new variable later. If something isn't obvious, please add a comment explaining what's going on.

4) Finally, related to the prior comment, the PEP8 guidelines officially recommend limiting lines to 79 characters. We haven't been super consistent with this, but we should try to avoid single super-long lines. You can look earlier in the code for examples of how we've done this (for example, things between parentheses can be split among two lines without an error). One reason it's important to prevent over-long lines is that if your screen isn't sufficiently wide, the line will either not be visible in its entirety, or it will wrap around and make it very difficult to decipher. This is especially true if you use split screens to, for example, compare code side-by-side.


Sorry for the very detailed (and possibly pedantic) comments, but all of this will a) prevent unexpected issues later, and b) make our code cleaner

File: src/book.py (hunk 1 above)
Comment (kdettman) [inline src/book.py:89]: Is there a function inside BookText that uses, or will use, this function? Methods that start with underscores are by convention only meant for internal use.

File: src/book.py [hunk 5]
```diff
@@ -82,7 +82,37 @@ def __init__(self, filepath=None, rawtext=None, encoding='utf-8', file_format='s
...
+        author1 = self._author
+        author2 = other._author
+        title1 = self._title
+        title2 = other._title
+        if (author1 is None and author2 is not None):
+            author = author2
+        elif (author1 is not None and author2 is None):
+            author = author1
+        elif (author1 != author2):
+            print ("The authors are not the same. Assigning the author of the 1st Book to the resultant bookobject")
```
Comment (kdettman) [inline src/book.py]: Change this to a warning.

File: src/book.py [hunk 6]
```diff
@@ -82,7 +82,37 @@ def __init__(self, filepath=None, rawtext=None, encoding='utf-8', file_format='s
...
+            author = author1
+        else:
+            author = author1
+            
+        if (title1 is None and title2 is not None):
+            title = title2
+        elif (title1 is not None and title2 is None):
+            title = title1
+        elif (title1 != title2):
+            print ("The authors are not the same. Assigning the title of the 1st Book to the resultant bookobject")
```
Comment (kdettman) [inline src/book.py]: Change this to a warning.

Comment (kdettman) [reply src/book.py]: Also, this mentions the authors rather than the titles.

File: src/book.py [hunk 7]
```diff
@@ -101,6 +131,7 @@ def clean(self, lemmatize=True, deromanize=False, lemma_pos='v', inplace=False):
         garbage = '\ufeff|â€™|â€"|â€œ|â€˜|â€\x9d|â€œi|_|â€'
         cleaned = re.sub(garbage, '', cleaned)
         cleaned = cleaned.replace('-', ' ')
+        cleaned = re.sub(r'\n+', ' ', cleaned)
```
Comment (kdettman) [inline src/book.py:136]: Should we just make this all string characters, something akin to regex '/s'?

File: src/book.py (hunk 1 above)
Comment (diptanil-roy) [reply src/book.py:89]: This is also the only way I know how to overload the '+' operator. For reference, https://www.geeksforgeeks.org/operator-overloading-in-python/

File: src/book.py (hunk 5 above)
Comment (diptanil-roy) [reply src/book.py]: Ok.

File: src/book.py (hunk 6 above)
Comment (diptanil-roy) [reply src/book.py]: Ok.

File: src/book.py (hunk 7 above)
Comment (diptanil-roy) [reply src/book.py:136]: We can. I will have to see how it behaves.

File: src/book.py (hunk 4 above)
Comment (diptanil-roy) [reply src/book.py]: Yes, it works either way.

File: src/book.py (hunk 2 above)
Comment (diptanil-roy) [reply src/book.py]: Ok. This is better.

File: src/book.py (hunk 1 above)
Comment (diptanil-roy) [reply src/book.py:89]: Ok.

File: src/book.py (hunk 7 above)
Comment (emapple) [reply src/book.py:136]: I suggested something like this elsewhere in the code (Kyle, I assume you meant whitespace characters, not string characters)

File: src/book.py [hunk 8]
```diff
@@ -205,6 +241,54 @@ def snippet(self, length, on, non_cont=False, inplace=False):
         else:
             return BookText(rawtext=snippet, author=self.author,
                             title=self.title, meta=self.meta)
+        
+    def random_snippet(self, on, n_groups=1, n_on=1, start_point=1, randomized = True, with_replacement=False, rem_stopwords=False, random_seed=0):
```
Comment (kdettman) [inline src/book.py]: Can this just be added as part of the snippet function? I thought we talked about adding features to the base method with an eye on expandability.

File: src/book.py (hunk 7 above)
Comment (kdettman) [reply src/book.py:136]: Ah yes, thanks for the catch. Regex is hard during group meetings.

Comment (emapple) [reply src/book.py:136]: Regex is always hard

File: src/book.py (hunk 8 above)
Comment (emapple) [reply src/book.py]: @kdettman I commented on this in a substantial amount of detail in my review; you may want to read through my comments there and see if you agree/disagree

File: src/book.py (hunk 1 above)
Comment (kdettman) [reply src/book.py:89]: Oh I see. I was misunderstanding what it was doing. Very clever.

File: src/book.py (hunk 8 above)
Comment (kdettman) [reply src/book.py]: I also see that this was addressed in the more general comment by Elaad above.

File: src/book.py [hunk 9]
```diff
@@ -82,7 +83,37 @@ def __init__(self, filepath=None, rawtext=None, encoding='utf-8', file_format='s
                 else:
                     self._title = None
         else:
-            self._title = None
+            self._title = title #FIXME IN MASTER
```
Comment (emapple) [inline src/book.py]:
```suggestion
            self._title = title
```

File: src/book.py [hunk 10]
```diff
@@ -151,7 +188,7 @@ def tokenize(self, on, rem_stopwords=True, stopword_lang='english',
             token = self._text.translate(
                 str.maketrans('', '', string.punctuation + '”“’'))
         if 'word' in on.lower():
-            token = word_tokenize(token.lower())
+            token = word_tokenize(token) #Changed here from token.lower() to token
```
Comment (emapple) [inline src/book.py:192]: Why did you change this?

File: src/book.py [hunk 11]
```diff
@@ -180,39 +217,79 @@ def tokenize(self, on, rem_stopwords=True, stopword_lang='english',
                     str.maketrans('', '', string.punctuation + '”“’'))
         return token
 
-    def snippet(self, length, on, non_cont=False, inplace=False):
+    def snippet(self, length, on, groups=1, non_cont=False, randomized=False, rem_stopwords=False, ret_as_arr=False, with_replace=True, random_seed=0, inplace=False):
```
Comment (kdettman) [inline src/book.py]: random_seed's default should be set to `None` rather than to 0 as 0 is a valid seed. Using `None` also aligns with what the random.seed() method uses as its default.

File: src/book.py [hunk 12]
```diff
@@ -180,39 +217,79 @@ def tokenize(self, on, rem_stopwords=True, stopword_lang='english',
...
+        rem_stopwords: Same as in Tokenize
+        ret_as_array: (Default = False): Returns one BookText object by appending all groups of snippets, else returns
+                                         an array of BookText objects
+        with_replace: (Default = True): For non-contiguous case, makes sure there is no repetition. If turned off, different
+                                        snippets in a group might have repeated char/word/sent.
+        random_seed: When positive, generates the same snippets everytime
+        
         """
-        # TODO add ability for random seed
+        if (random_seed != 0):
```
Comment (kdettman) [inline src/book.py]: See comment above, change to `if (random_seed is not None):`

File: src/book.py (hunk 11 above)
Comment (kdettman) [inline src/book.py]: The default value of `randomized` should be true since this was the previous default behavior.

File: src/book.py [hunk 13]
```diff
@@ -180,39 +217,79 @@ def tokenize(self, on, rem_stopwords=True, stopword_lang='english',
...
+                               author = self.author, title = self.title, meta = self.meta)
+                return_array.append(snippet)
+        else:
+            for gr in range(groups):
+                random_sample_of_indices = sorted(random.sample(range(0, len(tokens)), length), reverse=False)
+                dummy_string = str()
+                for num in random_sample_of_indices:
+                    dummy_string += tokens[num] + str(' ')
+                dummy_string.rstrip()    
+                snippet = BookText(rawtext= dummy_string, author = self.author,
```
Comment (kdettman) [inline src/book.py]: I believe for the sake of pep8 compliance the spaces between all `=` should be removed. This holds true for all of the calls of `BookText`.

File: src/book.py (hunk 11 above)
Comment (diptanil-roy) [reply src/book.py]: Ok.

File: src/book.py (hunk 13 above)
Comment (emapple) [reply src/book.py]: To clarify, only the spaces between `=` for arguments within function calls

File: src/book.py (hunk 12 above)
Comment (diptanil-roy) [reply src/book.py]: Ok.

File: src/book.py (hunk 11 above)
Comment (diptanil-roy) [reply src/book.py]: Ok.

File: src/book.py (hunk 13 above)
Comment (diptanil-roy) [reply src/book.py]: Ok. Thanks. 'Twas an oversight.

**Review by NikhilTilak — APPROVED** [review]
Looks good to me! I will add a couple changes to deromanize as a new pull.

File: src/book.py [hunk 14]
```diff
@@ -180,39 +217,79 @@ def tokenize(self, on, rem_stopwords=True, stopword_lang='english',
...
+            start = 0
+        
+        if non_cont == False:
+            for gr in range(groups):
+                snippet = BookText(rawtext=''.join((" ").join(tokens[length*gr:length*(gr+1)])), 
+                               author = self.author, title = self.title, meta = self.meta)
+                return_array.append(snippet)
+        else:
+            for gr in range(groups):
+                random_sample_of_indices = sorted(random.sample(range(0, len(tokens)), length), reverse=False)
```
Comment (kdettman) [inline src/book.py]: `random.sample(range(0, len(tokens)), length)` should be changed to `random.sample(range(0, len(tokens)) - length, length)` to prevent hitting the end of the book.

**Review by kdettman — CHANGES_REQUESTED** [review]
Some of the default behavior for the snippet function was changed which needs to be addressed for legacy sake. There is also a possible issue regarding possible overflow of a random snippet in line 263 that should either be changed or tested to make sure it doesn't break anything. Besides some additional minor formatting issues, this is a welcome upgrade to the class that both adds features while streamlining some of the features that already existed.

**Review by emapple — CHANGES_REQUESTED** [review]
In addition to Kyle's comments and the ones I've left elsewhere, I again have broader comments about `snippet`.

I'm not sure this has the behavior I expect; it's closer, but...

As it stands now, what is the purpose of `groups`? All it does is effectively act as a multiplier for `length`. There is no way to have different behavior _within_ vs. _between_ groups. It's either all consecutive or it's all random.

What I would expect from `non_cont=False, randomized=True` is you can have random starting points for contiguous chunks of text, and multiple groups would be separate contiguous chunks of text with random starting points.

In other words, there are really 4 combinations of random you can have:
1) random order within snippets, random order of snippets
2) random order within snippets, continuous order of snippets (not sure when this would ever be wanted)
3) contiguous order within snippets, random order of snippets (this would be my preferred default)
4) contiguous order within snippets, contiguous order of snippets (this is no different than just getting a single longer snippet and splitting it)

As I understand, right now only choices 1 and 4 are possible, which seems to make `groups` a redundant feature.

On top of this, we have choices for where to start (though to me, this only makes sense if we're not selecting randomly). We also have a choice of with_replacement, which doesn't seem to do what I expect.

File: src/book.py (hunk 10 above)
Comment (diptanil-roy) [reply src/book.py:192]: That was for my own sake. Part-of-speech tags rely on capitalisation at the beginning of sentences to identify.

Comment (emapple) [reply src/book.py:192]: Okay

File: src/book.py (hunk 14 above)
Comment (diptanil-roy) [reply src/book.py]: random.sample(range(0, len(tokens)), length) generates length number of integers between 0 and len(tokens). It will not hit the end of the book.

**Comment by diptanil-roy** [comment]
> In addition to Kyle's comments and the ones I've left elsewhere, I again have broader comments about `snippet`.
> 
> I'm not sure this has the behavior I expect; it's closer, but...
> 
> As it stands now, what is the purpose of `groups`? All it does is effectively act as a multiplier for `length`. There is no way to have different behavior _within_ vs. _between_ groups. It's either all consecutive or it's all random.

I need multiple snippets from a book, and not as one BookText objects as I am treating them separately. That is why the groups feature is there. If set to higher than 2, it returns multiple snippets of the same length from the text.

> What I would expect from `non_cont=False, randomized=True` is you can have random starting points for contiguous chunks of text, and multiple groups would be separate contiguous chunks of text with random starting points.

For non_cont=False, randomized=True, you will have a random starting point. But then, you will get a snippet where char/word/sent are contiguous. If you have groups = 2, you will get another snippet starting after the point the first snippet ended.

> In other words, there are really 4 combinations of random you can have:

This is not quite right. You never get a random order of snippets. They are always returned in the order they appeared in the book. Randomized is a random starting point, which is a legacy feature and was also pointed out by you in your comment above. 
So the 4 possibilities are:
1. Random contiguous: Random starting point. groups # of contiguous snippets, each of length.
Each snippet starts after the end point of the previous snippet.
2. Random non-contiguous: Random starting point. groups # of snippets, where the snippets are made of randomly picked char/word/sent after the starting point. If with_replacement=False, the same char/word/sent will not appear in 2 different snippets.
3. Non-random contiguous: Starting point is 0, groups # of contiguous snippets, each of length.
Each snippet starts after the end point of the previous snippet.
4. Non-random non-contiguous: Starting point is 0, groups # of snippets, where the snippets are made of randomly picked char/word/sent after the starting point. If with_replacement=False, the same char/word/sent will not appear in 2 different snippets.

> 1. random order within snippets, random order of snippets
> 2. random order within snippets, continuous order of snippets (not sure when this would ever be wanted)
> 3. contiguous order within snippets, random order of snippets (this would be my preferred default)
> 4. contiguous order within snippets, contiguous order of snippets (this is no different than just getting a single longer snippet and splitting it)
> 
> As I understand, right now only choices 1 and 4 are possible, which seems to make `groups` a redundant feature. 

Just to add here, groups is supposed to have a bit of redundancy, as instead of making groups more than 1, you could just set the value of length. However, when one requires a set of snippets from the same text (as I do), one needs the group feature. Even then, the ret_as_arr=True has to be set, so that you get back an array of BookText objects.

> On top of this, we have choices for where to start (though to me, this only makes sense if we're not selecting randomly). We also have a choice of with_replacement, which doesn't seem to do what I expect.

The choice of where to start is a legacy feature. You really do not have a choice of starting point. It's either randomly chosen or 0.

With_replacement only makes sense if you are using non-cont=True and groups>1. For non-contiguous text, with_replacement=False makes sure that the same char/word/sent doesn't show up in two different snippets. If you set it to True, the same char/word/sent can show up in  two different snippets.

File: src/book.py (hunk 7 above)
Comment (diptanil-roy) [reply src/book.py:136]: Can someone else please look into removing all string characters? For now, I want to keep it this way, and maybe address '\s' in a new pull request.

File: src/book.py (hunk 9 above)
Comment (diptanil-roy) [reply src/book.py]: Ok.

File: src/book.py (hunk 7 above)
Comment (emapple) [reply src/book.py:136]: This is already included in line 154 (or is automatically done if lemmatizing), so I don't think more work is needed

Comment (diptanil-roy) [reply src/book.py:136]: Ok. I am resolving this conversation then.

File: src/book.py (hunk 14 above)
Comment (diptanil-roy) [reply src/book.py]: len(tokens) is being updated at every step. That's why it will not hit the end of the book.

**Review by NikhilTilak — APPROVED** [review]

File: src/book.py [hunk 15]
```diff
@@ -180,39 +217,79 @@ def tokenize(self, on, rem_stopwords=True, stopword_lang='english',
...
+                               author = self.author, title = self.title, meta = self.meta)
+                return_array.append(snippet)
+        else:
+            for gr in range(groups): 
+                snippet = BookText(rawtext=''.join((" ").join(tokens[start:start+length])), 
+                               author = self.author, title = self.title, meta = self.meta)
+                snippet._text.rstrip()    
+                return_array.append(snippet)
+                
+                for index in sorted(range(start, start+length), reverse=True):
```
Comment (kdettman) [inline src/book.py:268]: Shouldn't this be from start to the length of the text -  (length + start)? That way it spans the whole space of the text rather than just from the starting position to the length of the snippet.

File: src/book.py [hunk 16]
```diff
@@ -180,39 +217,79 @@ def tokenize(self, on, rem_stopwords=True, stopword_lang='english',
...
+                    if (len(tokens)-length > 0):
```
Comment (kdettman) [inline src/book.py:269]: I'm unclear what this line is doing. Is tokens being updated at some point in this loop?

File: src/book.py [hunk 17]
```diff
@@ -180,39 +217,79 @@ def tokenize(self, on, rem_stopwords=True, stopword_lang='english',
...
         if inplace:
-            self._text = snippet
+            self._text = snip._text
         else:
-            return BookText(rawtext=snippet, author=self.author,
-                            title=self.title, meta=self.meta)
+            if ret_as_arr == False:
+                return snip
+            else:
+                return return_array
```
Comment (kdettman) [inline src/book.py]: Strictly speaking `return_array` isn't an array at least in the way I think of it. It is a list object. You could either use `np.asarray(return_array)` here or instantiate `return_array` as an empty array of length `groups`.

File: src/book.py (hunk 16 above)
Comment (diptanil-roy) [reply src/book.py:269]: Yes, to avoid repetition between snippets in a group of snippets, I drop the char/word/sent which already appeared.

File: src/book.py (hunk 17 above)
Comment (diptanil-roy) [reply src/book.py]: return_array was instantiated in the body.

File: src/book.py [hunk 18]
```diff
@@ -180,39 +217,79 @@ def tokenize(self, on, rem_stopwords=True, stopword_lang='english',
...
-            pass
+                "Argument 'on' must refer to character, word, or sentence")
+            
+        assert groups*length <= len(tokens)
+        return_array = []
+
+        if randomized:
+            start = random.randint(0, len(tokens) - groups*length)
+            tokens = tokens[start:]
+            start = 0
```
Comment (kdettman) [inline src/book.py:252]: `start` is set back to 0 here, when two lines above it is set as a random integer.

**Review by kdettman — CHANGES_REQUESTED** [review]
I have made some comments regarding some minor potential bugs and also some clarity in the function of the method.

File: src/book.py (hunk 15 above)
Comment (diptanil-roy) [reply src/book.py:268]: No, this is just to drop the char/word/sent that have already appeared in one snippet. There are length number of those only.

File: src/book.py (hunk 18 above)
Comment (diptanil-roy) [reply src/book.py:252]: because I am deleting everything in tokens upto the starting point. Therefore, start now coincides with the 1st element of tokens.

File: src/book.py (hunk 17 above)
Comment (kdettman) [reply src/book.py]: Yes but it is instantiated as `return_array = []` which is not a numpy array. This won't break anything, but it goes against expectations when the argument `ret_as_arr=True` gives you a list back.

**Comment by diptanil-roy** [comment]
> I have made some comments regarding some minor potential bugs and also some clarity in the function of the method.

I replied to each one of them above separately. Essentially, the idea is that I am updating the token object at every step, by deleting the char/word/sent that have already appeared. That helps me keep track sort of. There might be better ways to do it, but none of those were obvious to me.

File: src/book.py (hunk 16 above)
Comment (kdettman) [reply src/book.py:269]: But won't this give you either less groups than you asked for or a snippet which was not of the length requested?

Comment (kdettman) [reply src/book.py:269]: But won't `len(tokens)` always be greater than 0? So this step is essentially always going off? Haven't you already ensured that you won't get the exact same snippet twice by having your index set by `sorted(range(start, start+length)`?

File: src/book.py (hunk 17 above)
Comment (diptanil-roy) [reply src/book.py]: I see. Ok.

File: src/book.py (hunk 16 above)
Comment (diptanil-roy) [reply src/book.py:269]: it's len(tokens) - length. That changes at every iteration. This bit runs only when you have (say) 9 sentences, and you want 3 groups of 3 sentences each. In that case, this bit is satisfied at the last step.

File: src/book.py (hunk 17 above)
Comment (diptanil-roy) [reply src/book.py]: In the latest commit, it is returned as a numpy array.

**Review by emapple — CHANGES_REQUESTED** [review]
In testing this seems to behave as expected, but I can't follow your code very easily. Please comment it where it's not obvious what's happening.

File: src/book.py (hunk 18 above)
Comment (kdettman) [reply src/book.py:252]: I see

Comment (kdettman) [reply src/book.py:252]: Please add comments here so this is made clear

File: src/book.py (hunk 16 above)
Comment (kdettman) [reply src/book.py:269]: But to avoid repetition of snippets you only ever need to delete the first element in `tokens`. It should be ok if we ask for a 2 snippets and the first one is all of chapter 1 and the second one is all of chapter 1 minus the first word plus the first word of chapter 2.

**Comment by diptanil-roy** [comment]
> In testing this seems to behave as expected, but I can't follow your code very easily. Please comment it where it's not obvious what's happening.

I mentioned this to Kyle. If the behaviour is as expected, I do not want to spend more time writing comments now. I would like to work on my analysis too. I can revisit this to comment on it later.

**Comment by emapple** [comment]
> I mentioned this to Kyle. If the behaviour is as expected, I do not want to spend more time writing comments now. I would like to work on my analysis too. I can revisit this to comment on it later.

I _think_ it behaves as expected, based on a very simple test case.

But, as you wish. Just remember that you'll spend much longer finding unexplained bugs later on if it's not commented well and fully tested.

**Review by emapple — APPROVED** [review]

File: src/book.py (hunk 16 above)
Comment (kdettman) [reply src/book.py:269]: After in-person conversations, we were both looking for different implementations of this function. This disagreement is less important than getting this code out so I will add this as an "issue" to be resolved later.

**Comment by kdettman** [comment]
There is a slight disagreement between how I and the author think `snippet` should function, but those differences in opinion can be worked out later. For now I will approve these changes so that more work can be done on the project as a whole.

**Review by kdettman — APPROVED** [review]

**Comment by diptanil-roy** [comment]
> > I mentioned this to Kyle. If the behaviour is as expected, I do not want to spend more time writing comments now. I would like to work on my analysis too. I can revisit this to comment on it later.
> 
> I _think_ it behaves as expected, based on a very simple test case.
> 
> But, as you wish. Just remember that you'll spend much longer finding unexplained bugs later on if it's not commented well and fully tested.

Just to add a comment here: I have added comments on my local version, which is bare-bone but should be indicative enough for me to look at again if issues arise. 

Signing off!
